L85-core-binder-transport.md

```diff
@@ -0,0 +1,181 @@
+# Native cross-process communication on Android using BinderChannel
+
+*   Author(s): [Ming-Chuan Lin](https://github.com/sifmelcara),
+    [Ta-Wei Tu](https://github.com/taweitu),
+    [Nanping Jiang](https://github.com/napathome)
+*   Approver: TBD
+*   Status: Draft
+*   Implemented in: gRPC-core, C++
+*   Last updated: August 31, 2021
+*   Discussion at: https://groups.google.com/g/grpc-io/c/n8Qhlfwd8tE
+
+## Abstract
+
+Implements [BinderChannel protocol] (proposal L73) in gRPC-core transport layer, and provide
+C++ interface for Android native code.
+
+## Background
+
+Secure cross-process gRPC communication on Android was made possible by
+[BinderChannel protocol]. However, it is only supported in gRPC Java.
+
+For various reasons (performance, code reuse, etc) people need to write their
+code in C++. And sometimes they would like to create servers in the protocol or interact
+with servers implemented in the protocol.
+
+Although it is possible to use [JNI] to utilize gRPC Java's implementation of [BinderChannel protocol], it
+incurs significant overhead when JNI call is involved in every RPC call. Also
+makes the codebase harder to maintain.
+
+## Proposal
+
+We propose to implement [BinderChannel protocol] as a new transport in
+gRPC-Core, and expose C++ APIs along with some Java helper classes to support
+use scenario in Android native code.
+
+[NdkBinder] API will be used to implement the [wireformat] specified by the
+protocol.
+
+[JNI] and Java will be used by the transport implementation to set up the
+connection (APIs for setting up connections are only available in Java).
+After we establish the connection between server and client, the rest can be
+done in pure C++.
+
+### Client side API
+
+Since the channel creation process requires interactions with Android runtime,
+we will have a special API for creating binder transport channels.
+
+```cpp
+void CreateBinderChannel(
+    void* jni_env, jobject context, const std::string& package_name,
+    const std::string& class_name,
+    std::shared_ptr<grpc::binder::SecurityPolicy> security_policy,
+    std::function<void(void*, std::shared_ptr<grpc::Channel>)>
+        on_channel_created);
+
+class SecurityPolicy {
+ public:
+  // return true if the uid is authorized to connect
+  virtual bool IsAuthorized(void* jni_env, int uid) = 0;
```

Comment:
To my knowledge, there is no way to get access to PackageManager in NDK, so it would be easier on implementing teams if this function included the package name and signing certificate(s) to the c++ client; otherwise clients trying to implement security will need to include extra java code to call into PackageManager.

Alternatively, including a utility function to get the package name + signing certificate from c++ code could be used.

Reply:
Our plan is to provide many possible implementations of SecurityPolicy that makes sense, so client unlikely need to customize this

Yeah we can include a C++ utility function to get package name + signing certificate so that user can check certificate in pure C++. (This is a little bit off topic, and the utility function not necessary to be part of gRPC I guess)